decomposition.tex

```diff
@@ -0,0 +1,378 @@
+  return requires (Args&&...args) {
+    T{forward<Args>(args)...};
+  } &&
+  (is_reference<T>::value ||
+    (Destructible<T>() && requires (Args&&...args) {
+      new T{forward<Args>(args)...};
+    }));
+}
+\end{codeblock}
+
+\tcode{Constructible} describes either an object type \tcode{T} that is constructible from argument types
+\tcode{Args...}, or a reference type \tcode{T} that can be bound to \tcode{Args...}. For the object case,
+it includes allocation requirements per the design precept that ``construction is equivalent to allocation''
+and refines \tcode{Destructible} per ``destruction is prerequisite to construction.''
+
+\rSec2[decomposition.concepts.defaultconstructible]{DefaultConstructible (\ref{concepts.lib.object.defaultconstructible})}
+\begin{codeblock}
+template <class T>
+concept bool DefaultConstructible() {
+  return Constructible<T>() &&
+    requires (size_t n) {
+      new T[n]{};
+    };
+}
+\end{codeblock}
+
+\tcode{DefaultConstructible} adds an array allocation requirement onto the requirements of
+\tcode{Constructible<T>}. Since array initialization copy-initializes the array elements,
+this effectively requires both that any user-defined overload of \tcode{new T[]} is
+accessible, and that \tcode{T} not have an explicit default constructor.
+
+\rSec2[decomposition.concepts.moveconstructible]{MoveConstructible (\ref{concepts.lib.object.moveconstructible})}
+\begin{codeblock}
+template <class T>
+concept bool MoveConstructible() {
+  return Constructible<T, remove_cv_t<T>&&>() &&
+    requires (remove_cv_t<T> t) {
+      new T[1]{move(t)};
+    };
+}
+\end{codeblock}
+\tcode{MoveConstructible} describes an object type \tcode{T} whose objects can be move constructed:
+constructed from rvalues with the same value type (\tcode{remove_cv_t<T>}). It also requires array
+allocation, effectively forbidding \tcode{T} to have an explicit move constructor.
+
+\rSec2[decomposition.concepts.copyconstructible]{CopyConstructible (\ref{concepts.lib.object.copyconstructible})}
+\begin{codeblock}
+template <class T>
+concept bool CopyConstructible() {
+  return MoveConstructible<T>() &&
+    Constructible<T, remove_cv_t<T>&>() &&
+    Constructible<T, const remove_cv_t<T>&>() &&
+    Constructible<T, const remove_cv_t<T>&&>() &&
+    requires (remove_cv_t<T>& a, const remove_cv_t<T>& b) {
+      new T[1]{a};
+      new T[1]{b};
+      new T[1]{move(b)};
+    };
+}
+\end{codeblock}
+\tcode{CopyConstructible} describes an object type \tcode{T} whose objects can be copy constructed:
+constructed from (possibly \tcode{const}) lvalues or \tcode{const} rvalues with the same value type
+(\tcode{remove_cv_t<T>}). The array allocation requirements forbid \tcode{T} to have explicit
+copy constructor(s).
+
+\rSec2[decomposition.concepts.movable]{Movable (\ref{concepts.lib.object.movable})}
+\begin{codeblock}
+template <class T>
+concept bool Movable() {
+  return MoveConstructible<T>() &&
+    Assignable<T&, T&&>();
```

Comment:
If `T` is `const` or `const volatile` it makes no difference: both `Assignable<T&, T&&>` and `Assignable<T&, remove_cv_t<T>&&>` *should* be false. If `T` is just `volatile` then I have to admit I'm not sure if either choice is correct. I may want to assign the result of a function that returns `remove_cv_t<T>` to my `T` object, and/or I may want to assign one such object to another. Maybe the safe choice here is the strongest possible requirement: require both `Assignable<T&, T&&>` and `Assignable<T&, remove_cv_t<T>&&>` until/unless usage proves that is the wrong choice?